### PR TITLE
Fixes issue with sheet reset on rebuild

### LINKED ISSUE #101114: `DraggableScrollableSheet` doesn't maintain size after rebuild
## Steps to Reproduce

1. Run code below
2. Try expand the sheet by scroll, tap light and dark button
3. Minimize sheet, now expand it by tapping the button, tap light and dark button

**Expected results:** Sheet maintain it size

**Actual results:** Sheet lost the size and went back to initial size during rebuild if previous size was changed from controller

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();

  runApp(const App());
}

class App extends StatefulWidget {
  const App({Key? key}) : super(key: key);

  @override
  State<App> createState() => AppState();

  static AppState of(BuildContext context) {
    return context.findAncestorStateOfType<AppState>()!;
  }
}

class AppState extends State<App> {
  ThemeMode _themeMode = ThemeMode.light;

  void setThemeMode(ThemeMode mode) {
    if (_themeMode != mode) {
      setState(() {
        _themeMode = mode;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      themeMode: _themeMode,
      theme: ThemeData(brightness: Brightness.light),
      darkTheme: ThemeData(brightness: Brightness.dark),
      home: const Home(),
    );
  }
}

class Home extends StatelessWidget {
  const Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Demo'),
      ),
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            showModalBottomSheet(
              context: context,
              isScrollControlled: true,
              builder: (context) => const _DemoSheet(),
            );
          },
          child: const Text('Open'),
        ),
      ),
    );
  }
}

class _DemoSheet extends StatefulWidget {
  const _DemoSheet({Key? key}) : super(key: key);

  @override
  _DemoSheetState createState() => _DemoSheetState();
}

class _DemoSheetState extends State<_DemoSheet> {
  final _draggableController = DraggableScrollableController();

  void _expand() {
    _draggableController.animateTo(
      0.9,
      duration: const Duration(milliseconds: 150),
      curve: Curves.easeOut,
    );
  }

  void _setTheme(ThemeMode themeMode) {
    App.of(context).setThemeMode(themeMode);
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    return DraggableScrollableSheet(
      controller: _draggableController,
      initialChildSize: 0.5,
      minChildSize: 0.2,
      maxChildSize: 0.9,
      expand: false,
      builder: (context, scrollController) => ListView(
        physics: const BouncingScrollPhysics(),
        controller: scrollController,
        children: [
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: _expand,
              child: const Text('Expand'),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              'Current brightness: ${theme.brightness}',
              textAlign: TextAlign.center,
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: () => _setTheme(ThemeMode.light),
              child: const Text('Light'),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: ElevatedButton(
              onPressed: () => _setTheme(ThemeMode.dark),
              child: const Text('Dark'),
            ),
          ),
          for (int i = 0; i < 100; i++)
            Padding(
              padding: const EdgeInsets.all(8.0),
              child: Text(
                i.toString(),
                textAlign: TextAlign.center,
              ),
            ),
        ],
      ),
    );
  }
}
```

</details>

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel unknown, 2.10.4, on macOS 12.1 21C52 darwin-arm, locale en-ID)
    • Flutter version 2.10.4 at /Users/s21/Documents/GitHub/flutter/flutter
    • Upstream repository unknown
    • Framework revision c860cba910 (6 days ago), 2022-03-25 00:23:12 -0500
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/s21/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.66.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.36.0

[✓] Connected device (3 available)
    • iPhone SE (3rd generation) (mobile) • E31DBB95-A1F1-4893-AF0C-48C839A6A434 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-4 (simulator)
    • macOS (desktop)                     • macos                                • darwin-arm64   • macOS 12.1 21C52 darwin-arm
    • Chrome (web)                        • chrome                               • web-javascript • Google Chrome 99.0.4844.84

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

https://user-images.githubusercontent.com/1574722/161059139-890ee69f-9f3a-49ba-ad79-9b086fd54a0d.mov

### PR DESCRIPTION
**Pull request by caseycrogers**
There was a bug where the sheet would reset to `initialChildSize` on rebuild if the last change to the sheet was programmatic. This PR introduces a test and fix for this behavior.

Fixes: https://github.com/flutter/flutter/issues/101114

Having both `hasDragged` and `hasChanged` feels quite sloppy to me, but they're the only way to produce the developer-expected behavior here (see code comment above where the two are declared).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by caseycrogers** [comment]
@dnfield Do you mind looking at this one?

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart [hunk 1]
```diff
@@ -505,6 +509,16 @@ class _DraggableSheetExtent {
   // this because we don't want to snap away from an initial or programmatically set size.
   bool hasDragged;
 
+  // Used to determine if the sheet should move to the new initial size when it
+  // changes.
+  // We need both `hasChanged` and `hasDragged` to achieve the following
+  // behavior:
+  //   1. The sheet should only snap following user drags (as opposed to
+  //      programmatic sheet changes.
```
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:517]: Is this publicly documented somewhere?

Comment (dnfield) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:517]: I think this behavior probably makes sense but we should document it.

Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart]: done

File: packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart [hunk 2]
```diff
@@ -101,6 +101,10 @@ class DraggableScrollableController extends ChangeNotifier {
   /// The duration must not be zero. To jump to a particular value without an
   /// animation, use [jumpTo].
   ///
+  /// The sheet will not snap after calling [animateTo] even if it is not at a
+  /// snap point and [DraggableScrollableSheet.snap] is true. Snapping only
+  /// occurs after user drags.
```
Comment (dnfield) [inline packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart:106]: I think the `even if it not as a snap point` is redundant and can be removed. Maybe just

> The sheet will not snap after calling animateTo even if [DraggableScrollSheet.snap] is true. Snapping only occurs after user drags.

We should also document this on `DraggableScrollableSheet.snap` and maybe on the top level docs, e.g.

> The snap property only affects user drags. Programmatically manipulating the scroll offset via [DraggableScrollableController.animteTo] or [DraggableScrollableController.jumpTo] will ignore the [snap] property and the snap offsets.

Comment (caseycrogers) [reply packages/flutter/lib/src/widgets/draggable_scrollable_sheet.dart]: This all seems reasonable. Did the following:
1) Removed `even if it is not at a snap point`
2) Added `above snapping clarification (with minor edits) to class level docs
3) Added more brief version of snaping clarification to `snap` attribute docs

**Comment by goderbauer** [comment]
fly-by comment from triage: Looks like the doc check is unhappy.

**Comment by caseycrogers** [comment]
> fly-by comment from triage: Looks like the doc check is unhappy.

Thanks, fix'd! That's what I get for writing commits past midnight.

**Review by dnfield — CHANGES_REQUESTED** [review]
This looks like it's still missing docs on the `drag` property and on the class itself.

**Comment by caseycrogers** [comment]
> This looks like it's still missing docs on the `drag` property and on the class itself.

Hmm I might be misunderstanding something.

I've put comments on snapping property:

On the snap attribute
https://github.com/flutter/flutter/pull/107876/files#diff-8a2b7e9da235db9760f0046cbbceddc5e41c78c9285a315141c401f715bd269fR357

On the top level class
https://github.com/flutter/flutter/pull/107876/files#diff-8a2b7e9da235db9760f0046cbbceddc5e41c78c9285a315141c401f715bd269fR242

And on animateTo+JumpTo
https://github.com/flutter/flutter/pull/107876/files#diff-8a2b7e9da235db9760f0046cbbceddc5e41c78c9285a315141c401f715bd269fR104
https://github.com/flutter/flutter/pull/107876/files#diff-8a2b7e9da235db9760f0046cbbceddc5e41c78c9285a315141c401f715bd269fR156

Let me know if there's anywhere else you want me to add docs.

**Review by dnfield — APPROVED** [review]
LGMT - sorry,t he git diff collapsing threw me for a loop.

**Comment by dnfield** [comment]
@Piinks would you mind taking a second pass on this?

**Review by Piinks — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**